File: 3A/se382/A2/lsystem/DragonCurve.java
/*
 *
 */
package lsystem;

import java.awt.Rectangle;
import java.awt.geom.Point2D;
import java.awt.geom.Point2D.Double;
import java.util.Vector;

public class DragonCurve extends AbstractIteratedSystem {

    private TurtleGraphics turtle = new TurtleGraphics();
    Vector<Point2D.Double> points = new Vector<Point2D.Double>();
    String                 curString = "";
    
    public DragonCurve() {
        super(15);
        this.reset();
    }

    public String getName() {
        return "Dragon Curve";
    }

    public Vector<Double> getLines(Rectangle displayArea) {
        return this.scalePointsToArea(points, displayArea);
    }

    public void reset() {
        super.reset();
        curString = "FX";
        turtle.setLocation(new Point2D.Double(0,0));
        turtle.pointRight();
        points.clear();
        points.add(turtle.getLocation());
    }

    protected void iterateSystem() {
        StringBuffer newString = new StringBuffer();
        
        turtle.setLocation(new Point2D.Double(0,0));
        turtle.pointRight();
        points.clear();
        points.add(turtle.getLocation());
        
        for (int i = 0; i < curString.length(); i++) {
            switch (curString.charAt(i)) {
                case 'F':
                    points.add(turtle.moveForward());
                    newString.append(curString.charAt(i));
                    break;
                case 'X':
                    newString.append("X+YF+");
                    break;
                case 'Y':
                    newString.append("-FX-Y");
                    break;
                case '-':
                    turtle.turn(-90);
                    newString.append(curString.charAt(i));
                    break;
                case '+':
                    turtle.turn(90);
                    newString.append(curString.charAt(i));
                    break;
            }
        }
        curString = newString.toString();
    }
}

File: 3A/se382/A1/lsystem/DragonCurve.cpp
#include "DragonCurve.h"

#define MAX_ITERATIONS 15

namespace cs349
{

DragonCurve::DragonCurve() : IteratedSystem(MAX_ITERATIONS)
{
	this->Reset();
}

DragonCurve::~DragonCurve()
{
}

string DragonCurve::GetName() const
{
	return string("Dragon Curve");
}

void DragonCurve::SetParamValue(int paramNum, int value)
{
	; // no-op
}

void DragonCurve::Reset()
{
	IteratedSystem::Reset();
	curString = string("FX");
	this->turtle.SetLocation(Point(0,0));
	this->turtle.PointRight();
	this->points.clear();
	this->points.push_back(this->turtle.GetLocation());
}

void DragonCurve::IterateSystem()
{
	// Algorithm source: http://en.wikipedia.org/wiki/L-system
	string newString;

	this->turtle.SetLocation(Point(0,0));
	this->turtle.PointRight();
	this->points.clear();
	this->points.push_back(this->turtle.GetLocation());

	for (string::iterator iter=this->curString.begin(); iter != this->curString.end(); iter++) {
		char c = (*iter);
		if (c == 'F') {
			this->points.push_back(this->turtle.MoveForward());
			newString.push_back(c);
		} else if (c == 'X') {
			newString.append("X+YF+");
		} else if (c == 'Y') {
			newString.append("-FX-Y");
		} else if (c == '-') {
			this->turtle.Turn(-90);
			newString.push_back(c);
		} else if (c == '+') {
			this->turtle.Turn(90);
			newString.push_back(c);
		}
	}
	this->curString = newString;
}

vector<Point> DragonCurve::GetLines(const Rectangle & displayArea) const
{
	return this->ScalePointsToArea(this->points, displayArea);
}

}
